crates/core/src/subscription/delta.rs

```diff
@@ -21,24 +23,93 @@ pub fn eval_delta<'a, Tx: Datastore + DeltaStore>(
     plan: &SubscriptionPlan,
 ) -> Result<Option<UpdatesRelValue<'a>>> {
     metrics.delta_queries_evaluated += 1;
+
     let mut inserts = vec![];
     let mut deletes = vec![];
 
-    plan.for_each_insert(tx, metrics, &mut |row| {
-        inserts.push(row.into());
-        Ok(())
-    })?;
+    let mut duplicate_rows_evaluated = 0;
+    let mut duplicate_rows_sent = 0;
+
+    // Query plans for joins may return redundant rows,
+    // but we track row counts to avoid sending them to clients.
+    //
+    // Single table plans will never return redundant rows,
+    // so there's no need to track row counts.
```

Comment:
This comment is confusing. You take the `else` part and in fact update the counts.